tests/mpi/distribute_flux_sparsity_pattern.cc

```diff
@@ -0,0 +1,323 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+/*
+ * Verify that the current commit fixes distributed flux sparsity
+ * patterns. This fixes a bug where (see the commit for more details) coupling
+ * between an active, locally owned cell and an active ghost cell may be
+ * erroneously skipped when building a sparsity pattern with flux terms.
```

Comment:
One question regarding the test and the change you made in dof_tools: Is the error because of your strategy to select which processor writes into the matrix that is done differently in your code as compared to the creation of the sparsity pattern. Or do we indeed miss terms in the final sparse matrix (as shown by the number of nonzeros in the matrix)? I agree that the above code fixes the problem by being more conservative about which processor will process a face at a processor boundary, simply by adding terms in both ways.

Or put differently: If you used the same pattern for which processor should process a cell in the assembly as were used in the **old** version of the flux sparsity pattern method, would the error go away? Or would you miss some face terms by that algorithm because the `cell->index()` is gone through in a different way on different processors?